ENIGMAsystem/SHELL/Universal_System/Extensions/Json/short_json_converter.cpp

```diff
@@ -0,0 +1,262 @@
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License
+ *      along with this program.  If not, see <http://www.gnu.org/licenses/>.
+**/
+
+#include "short_json.h"
+
+namespace enigma {
+class ShortJSONReader {
+ private:
+  enum ShortJSONState { arrayState, objectState, valueState, successState, errorState };
+  int pointer_{-1};
+  ShortJSONState currentState;
+  std::stack<std::pair<std::string, bool>> levels;
+  std::stack<int> indices;
+
+  ShortJSONState pushLevel(std::string data) {
+    while (1) {
+      pointer_++;
+      if (!(pointer_ < ((int)(data.length())))) return successState;
+      switch (data.at(pointer_)) {
+        case '[':
+          levels.push({"", true});  // we assume worst case which is array of objects until otherwise
+          indices.push(0);
+          currentState = pushLevel(data);
+          switch (currentState) {
+            case errorState:
+              return errorState;
+            default:
+              break;
+          }
+          continue;
+        case ']':
+          indices.pop();
+          currentState = popLevel();
+          switch (currentState) {
+            case errorState:
+              return errorState;
+            default:
+              break;
+          }
+          continue;
+        case '"':
+        case '0':
+        case '1':
+        case '2':
+        case '3':
+        case '4':
+        case '5':
+        case '6':
+        case '7':
+        case '8':
+        case '9':
+        case '-':
+        case 't':
+        case 'f':
+          levels.top().second = false;  // then it's object
+          currentState = read_value(data);
+          switch (currentState) {
+            case errorState:
+              return errorState;
+            default:
+              break;
+          }
+          continue;
+        case ',':
+          levels.top().first += ',';
+          indices.top()++;
+          continue;
+        default:
+          return errorState;
+      }
+    }
+  }  // pushLevel
+
+  ShortJSONState popLevel() {
+    if (levels.top().second)
+      return array_to_json(levels.top().first);
+    else
+      return object_to_json(levels.top().first);
+  }  // popLevel()
+
+  ShortJSONState read_value(std::string data) {
+    switch (data.at(pointer_)) {
+      case '"':
+      case '0':
+      case '1':
+      case '2':
+      case '3':
+      case '4':
+      case '5':
+      case '6':
+      case '7':
+      case '8':
+      case '9':
+      case '-':
+      case 't':
+      case 'f': {
+        int temp_pointer_{pointer_};
+        while (((temp_pointer_ + 1) < ((int)(data.length()))) && ((data.at(temp_pointer_ + 1)) != ',') &&
+               ((data.at(temp_pointer_ + 1) != ']')))
+          temp_pointer_++;
+        if (!(temp_pointer_ < ((int)(data.length())))) return errorState;
```

Comment:
Same thing here as before, this can be simplified to `temp_pointer_ >= data.length()`.